test/temporal.jl

```diff
@@ -0,0 +1,312 @@
+
+    hod = HoD()
+    @test hod isa Transform
+
+    @testset "Vector" begin
+        x = collect(DateTime(2020, 1, 1, 9, 0):Hour(1):DateTime(2020, 5, 7, 9, 0))
+        # Expected result is an hour a day starting and ending on the 9th hour inclusive,
+        # with 126 full days in the middle
+        expected = [9:23..., repeat(0:23, 126)..., 0:9...]
+
+        @test Transforms.apply(x, hod) == expected
+        @test hod(x) == expected
+
+        # Test the tranform was not mutating
+        @test x != expected
+
+        @testset "StepRange" begin
+            x = DateTime(2020, 1, 1, 9, 0):Hour(1):DateTime(2020, 5, 7, 9, 0)
+            @test Transforms.apply(x, hod) == expected
+            @test hod(x) == expected
+        end
+
+        @testset "DST" begin
+            x = ZonedDateTime(2020, 3, 7, 9, 0, tz"America/New_York"):Hour(1):ZonedDateTime(2020, 3, 8, 9, 0, tz"America/New_York")
+
+            # expected result skips the DST transition hour of 2
+            expected = [9:23..., 0, 1, 3:9...]
+
+            @test Transforms.apply(x, hod) == expected
+            @test hod(x) == expected
+        end
+
+        @testset "dims" begin
+            @testset "dims = :" begin
+                d = Colon()
+                @test Transforms.apply(x, hod; dims=d) == expected
+                @test hod(x; dims=d) == expected
+            end
+
+            @testset "dims = 1" begin
+                d = 1
+                expected = [[x] for x in [9:23..., 0, 1, 3:9...]]
+                @test Transforms.apply(x, hod; dims=d) == expected
+                @test hod(x; dims=d) == expected
+            end
+        end
+
+        @testset "inds" begin
+            expected = collect(10:13)
+            @test Transforms.apply(x, hod; inds=2:5) == expected
+            @test hod(x; inds=2:5) == expected
+
+            @testset "dims = 1" begin
+                d = 1
+                expected = [[x] for x in [9:23..., 0, 1, 3:9...]]
+                @test Transforms.apply(x, hod; dims=d, inds=[1]) == expected
+            end
+        end
+    end
+
+    @testset "Matrix" begin
+        x = [
+            DateTime(2020, 1, 1, 1, 0):Hour(1):DateTime(2020, 1, 1, 3, 0);
+            DateTime(2020, 1, 1, 9, 0):Hour(1):DateTime(2020, 1, 1, 11, 0)
+        ]
+        M = reshape(x, 3, 2)
+        expected = [1 9; 2 10; 3 11]
+
+        @test Transforms.apply(M, hod) == expected
+        @test hod(M) == expected
+
+        # Test the tranform was not mutating
+        @test M != expected
+
+        @testset "dims" begin
+            @testset "dims = :" begin
+                d = Colon()
+                @test Transforms.apply(M, hod; dims=d) == expected
+                @test hod(M; dims=d) == expected
+
+                # Test the tranform was not mutating
+                @test M != expected
+            end
+
+            @testset "dims = 1" begin
+                d = 1
+                expected = [[1, 9], [2, 10], [3, 11]]
```

Comment:
I guess this is an example of a return being different to it's input? Matrix vs Vector{Vector}?

Reply:
We could `cat` the result and keep the input/output consistent, but `eachslice` doesn't preserve orientation, e.g.

```julia
julia> A = [1 2 3; 4 5 6; 7 8 9];

julia> r = [x .+ 1 for x in eachslice(A, dims=1)]
3-element Array{Array{Int64,1},1}:
 [2, 3, 4]
 [5, 6, 7]
 [8, 9, 10]

julia> s = [x .+ 1 for x in eachslice(A, dims=2)]
3-element Array{Array{Int64,1},1}:
 [2, 5, 8]
 [3, 6, 9]
 [4, 7, 10]
```

we could instead replace it with `mapslices` for which 
>The results are concatenated along the remaining dimensions

```julia
julia> t = mapslices(x->x.+1, A, dims=1)
3×3 Array{Int64,2}:
 2  3   4
 5  6   7
 8  9  10

julia> t = mapslices(x->x.+1, A, dims=2)
3×3 Array{Int64,2}:
 2  3   4
 5  6   7
 8  9  10
```

Reply:
I think doing something like this would work
```julia
julia> function apply(A::AbstractArray, t::HoD; dims=:, inds=:, kwargs...)
           if dims === Colon()
               if inds === Colon()
                   return _apply(A, t; kwargs...)
               else
                   return _apply(A[inds], t; kwargs...)
               end
           end

           return mapslices(x->_apply(x[inds], t; kwargs...), A, dims=dims)
       end
apply (generic function with 1 method)

julia>

julia>

julia> apply(M, HoD(), dims=1)
3×2 Array{Int64,2}:
 1   9
 2  10
 3  11

julia> apply(M, HoD(), dims=2)
3×2 Array{Int64,2}:
 1   9
 2  10
 3  11
```

Reply:
I made a different thread about this but if we `cat` the result, for this transform, why do we need `dims`? As per your example the result is the same. One reason is this method might become common and generalised in `transformers.jl`?

Reply:
probably we don't need dims then - and it should be absorbed in the `kwargs...`

Reply:
Yeah I was aiming on having this apply method be the default, and moving it to transforms.jl if it's cleared as being a good general case